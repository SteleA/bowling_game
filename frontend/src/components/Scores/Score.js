import React from 'react';
import _ from 'lodash';

const score = (props) => {
  const isExtra = props.rolls.length === 3;
  return (
    <div className={`score ${props.active ? 'active' : ''}`}>
      <span className="round">{props.round}</span>
      <div className={`score-body ${isExtra ? 'extra-roll' : ''}`}>
        {_.map(props.rolls, roll =>
          <span
            key={_.uniqueId()}
            className="score-box"
          >{roll.score}</span>)}
        <div className="currentScore">{props.score}</div>
      </div>
    </div>
  );
};

export default score;
