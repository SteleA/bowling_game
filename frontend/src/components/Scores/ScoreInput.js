import React from 'react';
import { FormGroup, FormControl } from 'react-bootstrap';

export default props => (
  <FormGroup>
    <FormControl
      id={props.id}
      type="number"
      value={props.input}
      onChange={({ target: { value } }) => props.onChange(value)}
      placeholder={props.placeholder}
    />
  </FormGroup>
);
