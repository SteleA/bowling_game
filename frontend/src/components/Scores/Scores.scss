.scores {
  $border-color: #bebebe;
  $inactive-bg: #ececec;
  $active-border: red;
  $active-bg: white;

  margin-bottom: 25px;
  .score {
    &.active {
      .score-body {
        border: 1px solid $active-border;
        background: $active-bg;
        .score-box {
          border-top: 0px;
          &:first-of-type {
            border-left: 0px
          }
          &:last-of-type {
            border-right: 0px
          }
        }
      }
    }
    width: calc(10% - 10px);
    margin: 0 5px;
    float: left;
    text-align: center;
    .round {
      font-weight: 600;
    }
    .score-body {
      &.extra-roll {
        .score-box {
          width: calc(100% / 3);
          &:nth-of-type(2) {
            position: relative;
            display: inline-block;
            border-right: 0px;
          }
        }
        .currentScore {
          padding-top: 10px;
        }
      }
      background: $inactive-bg;
      height: 100px;
      position: relative;
      .score-box {
        &:first-of-type {
          left: 0;
          border-left: 0px;
        }
        &:last-of-type {
          right: 0;
        }
        height: 40px;
        position: absolute;
        padding: 10px;
        border-bottom: 1px solid $border-color;
        border-left: 1px solid $border-color;
        width: 50%;
      }
      .currentScore {
        padding-top: 50px;
        font-size: 24px;
      }
    }
  }
}
