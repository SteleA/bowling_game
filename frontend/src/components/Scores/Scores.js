import React from 'react';
import {Row, Col} from 'react-bootstrap';
import _ from 'lodash';
import Score from './Score';
import './Scores.scss';

const scores = (props) => {
  return (
    <Row className="scores">
      {_.map(props.scores, score =>
        <Score key={_.uniqueId()} {...score} active={score.round === props.round} />)}
    </Row>
  );
};

export default scores;
