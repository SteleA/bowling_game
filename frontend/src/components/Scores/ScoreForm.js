import React, {Component} from 'react';
import _ from 'lodash';
import {Button} from 'react-bootstrap';
import ScoreInput from './ScoreInput';

class ScoreForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      scores: {},
    };
  }

  onAddScores(e) {
    if (e) e.preventDefault();
    const rolls = _.map(this.props.round.rolls, roll =>
      ({
        ...roll,
        score: this.getNumber(this.state.scores[roll.id]),
      }));

    const round =
      Object.assign({},
        this.props.round,
        {rolls, score: this.calculateScores(this.state.scores)});

    this.props.addScore(round);
    this.setState({scores: {}});
  }

  getNumber(num) {
    const number = _.toNumber(num);
    return _.isNaN(number) ? 0 : number;
  }

  setScore(id, score) {
    const scores = Object.assign({}, this.state.scores);
    scores[id] = score;
    const totalScore = this.calculateScores(scores) <= 10;
    const lastRound = this.calculateScores(scores) <= 30 && this.props.round.round === 10;
    if ((totalScore || lastRound) && score >= 0) {
      this.setState({scores});
    }
    return false;
  }

  calculateScores(scores) {
    return _.chain(scores)
        .map(score => score)
        .map(_.toNumber)
        .reduce((acc, cur) => acc + cur)
        .value();
  }

  render() {
    return (
      <div>
        {_.map(this.props.round.rolls, roll =>
          <ScoreInput
            id={`score-input-${roll.id}`}
            key={`score-input-${roll.id}`}
            onChange={score => this.setScore(roll.id, score)}
            input={this.getNumber(this.state.scores[roll.id])}
            placeholder={roll.name}
          />
        )}
        <Button block onClick={() => this.onAddScores()}>Submit Scores</Button>
      </div>
    );
  }
}

export default ScoreForm;
