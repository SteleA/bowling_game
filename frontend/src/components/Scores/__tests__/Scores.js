import { shallow } from 'enzyme';  // eslint-disable-line
import { shallowToJson } from 'enzyme-to-json';  // eslint-disable-line
import React from 'react';
import Scores from '../Scores';

const props = {};

describe('<Scores />', () => {
  it('should render with default props', () => {
    const wrapper = shallow(
      <Scores {...props} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
