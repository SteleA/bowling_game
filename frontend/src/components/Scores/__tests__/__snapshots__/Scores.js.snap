// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Scores /> should render with default props 1`] = `
<Row
  bsClass="row"
  className="scores"
  componentClass="div"
/>
`;
